core/src/main/java/org/mule/runtime/core/internal/streaming/StreamingGhostBuster.java

```diff
@@ -0,0 +1,164 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+
+/**
+ * Tracks instances of {@link ManagedCursorProvider} through the {@link #track(ManagedCursorProvider)} method.
+ * This class uses a {@link ReferenceQueue} so that when each of those instances are garbage collected, we can make sure
+ * that {@link ManagedCursorProvider#releaseResources()} is invoked.
+ * <p>
+ * This is useful in cases of long running flows in which cursor providers are open and dereferenced long before the flow ends
+ * (e.g: an &lt;until-successful&gt; which reads a file many times or a &lt;foreach&gt; that executes many queries which result
+ * are never used outside of the scope).
+ * <p>
+ * Notice that this <b>DOES NOT</b> replace the cleanup and tracking job that is performed in {@link CursorManager}. That
+ * still remains the main cleanup mechanism as we cannot rely on garbage collection for releasing streaming resources. This is
+ * only a mitigation for cases like described above.
+ *
+ * @since 4.2.0
+ */
+public class StreamingGhostBuster implements Lifecycle {
+
+  private static final long POLL_INTERVAL = SECONDS.toMillis(5);
+  private static final Logger LOGGER = getLogger(StreamingGhostBuster.class);
+
+  private final ReferenceQueue<ManagedCursorProvider> referenceQueue = new ReferenceQueue<>();
+  private volatile boolean stopped = false;
+  private Future taskHandle;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private Scheduler scheduler;
+
+  @Override
+  public void initialise() throws InitialisationException {
+    scheduler = schedulerService.customScheduler(SchedulerConfig.config()
+        .withMaxConcurrentTasks(1)
+        .withName("StreamingManager-CursorProviderCollector"));
+  }
+
+  @Override
+  public void start() throws MuleException {
+    try {
+      taskHandle = scheduler.submit(this::bustGhosts);
+    } catch (RejectedExecutionException e) {
+      throw new MuleRuntimeException(e);
+    }
+    stopped = false;
+  }
+
+  @Override
+  public void stop() throws MuleException {
+    stopped = true;
+    taskHandle.cancel(true);
+    taskHandle = null;
+  }
+
+  @Override
+  public void dispose() {
+    scheduler.stop();
+  }
+
+  /**
+   * Tracks the given {@code cursorProvider}
+   *
+   * @param cursorProvider a {@link ManagedCursorProvider}
+   * @return a {@link WeakReference} wrapping the {@code cursorProvider}
+   */
+  public WeakReference<ManagedCursorProvider> track(ManagedCursorProvider cursorProvider) {
+    return new StreamingWeakReference(cursorProvider, referenceQueue);
+  }
+
+  private void bustGhosts() {
+    while (!stopped && !currentThread().isInterrupted()) {
+      try {
+        StreamingWeakReference ghost = (StreamingWeakReference) referenceQueue.remove(POLL_INTERVAL);
+        if (ghost != null) {
+          bust(ghost);
+        }
+      } catch (InterruptedException e) {
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Either re-interrupt this method or rethrow the "InterruptedException". ([squid:S2142](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS2142))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7648&issues=AWi-qz_DotIXnmsUpxIW&open=AWi-qz_DotIXnmsUpxIW)

[ISSUE_KEY]: AWi-qz_DotIXnmsUpxIW